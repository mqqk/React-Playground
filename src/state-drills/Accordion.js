import React from 'react';

export default class Accordion extends React.Component{

    static defaultProps = {
        sections:[]
    };
    
     state = {
       currentSectionIndex:null,
    }

    renderList(currentSectionIndex){
        console.log('running renderList')
        console.log(this.props.sections);
        return <ul> {this.renderListButtons(currentSectionIndex)}</ul>
    }

    renderListButtons(currentSectionIndex){
        console.log('running ListButtons')
        return this.props.sections.map((section,index) => (
            <li key={index}>
                <button type='button' key={index} onClick={() => this.buttonClicks(index)}> {section.title} </button>
                {(currentSectionIndex===index) && <p>{section.content}</p>}
            </li>
        ))
        
    }

    buttonClicks(index){
        console.log('buttonClicks running');
        this.setState({currentSectionIndex:index})
        console.log(this.state.currentSectionIndex);
    }


    render() {
         const { currentSectionIndex } = this.state       

        return(
        <div>
            {this.renderList(currentSectionIndex)}
            
        </div>
        )
    }



}


// The accordion renders a <ul>, inside the ul will be a li element for each section. Each section's li should be a button containing the section's title.
// When a button is clicked, a p should be rendered below the button that was clicked (inside the li) that displays the section's content. You can use conditional rendering to display this p with the section content.
// When a button is clicked, only that section should be open and the other sections should be closed.
// You should write at least 4 snapshot tests for the Accordion component:
// The component renders an empty li when the sections prop is not supplied.
// The component renders no sections as active by default.
// The component opens a clicked section.
// The component only opens the last section when multiple sections have been clicked.
// As a stretch goal, add styles for the Accordion component in an Accordion.css file.