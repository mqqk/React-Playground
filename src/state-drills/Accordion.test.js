import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Accordion from './Accordion';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Accordion Compnent',() =>{

    // The component renders an empty li when the sections prop is not supplied.
    // The component renders no sections as active by default.
    // The component opens a clicked section.
    // The component only opens the last section when multiple sections have been clicked.

    const sectionsProp = [
        {
          title:'Section 1',
          content:'First Section has anything but Latin'
        },
        {
          title:'Section 2',
          content:'This wil all make perfect sense later'
        },
        {
          title:'Section 3',
          content:'Keep it rolling'
        }
      
      ]

      it('renders an empty li when the section prop is not supplied', () => {
        const wrapper = shallow(<Accordion />)
        expect(toJson(wrapper)).toMatchSnapshot()

      })

      it('renders no section as active by defaul',() =>{
          const wrapper = renderer.create(<Accordion sections={sectionsProp}/>)
          expect(toJson(wrapper)).toMatchSnapshot()
      })

      it('component opens a clicked seaction',() =>{
          const wrapper = shallow(<Accordion sections={sectionsProp}/>)
          wrapper.find('button').at(1).simulate('click')
          expect(toJson(wrapper)).toMatchSnapshot()
      })

      it('component only opens last section when others have been clicked', () => {
          const wrapper = shallow(<Accordion sections={sectionsProp} />)
          wrapper.find('button').at(1).simulate('click')
          wrapper.find('button').at(2).simulate('click')
          expect(toJson(wrapper)).toMatchSnapshot()
      })




})