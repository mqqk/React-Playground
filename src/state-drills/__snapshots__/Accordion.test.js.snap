// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Compnent component only opens last section when others have been clicked 1`] = `
<div>
  <ul>
     
    <li
      key="0"
    >
      <button
        key="0"
        onClick={[Function]}
        type="button"
      >
         
        Section 1
         
      </button>
    </li>
    <li
      key="1"
    >
      <button
        key="1"
        onClick={[Function]}
        type="button"
      >
         
        Section 2
         
      </button>
    </li>
    <li
      key="2"
    >
      <button
        key="2"
        onClick={[Function]}
        type="button"
      >
         
        Section 3
         
      </button>
      <p>
        Keep it rolling
      </p>
    </li>
  </ul>
</div>
`;

exports[`Accordion Compnent component opens a clicked seaction 1`] = `
<div>
  <ul>
     
    <li
      key="0"
    >
      <button
        key="0"
        onClick={[Function]}
        type="button"
      >
         
        Section 1
         
      </button>
    </li>
    <li
      key="1"
    >
      <button
        key="1"
        onClick={[Function]}
        type="button"
      >
         
        Section 2
         
      </button>
      <p>
        This wil all make perfect sense later
      </p>
    </li>
    <li
      key="2"
    >
      <button
        key="2"
        onClick={[Function]}
        type="button"
      >
         
        Section 3
         
      </button>
    </li>
  </ul>
</div>
`;

exports[`Accordion Compnent renders an empty li when the section prop is not supplied 1`] = `
<div>
  <ul>
     
  </ul>
</div>
`;

exports[`Accordion Compnent renders no section as active by defaul 1`] = `undefined`;
